### PR TITLE
Initial conversion

### LINKED ISSUE #1: Initial conversion
We need to move the logic from SEER\*DMS to this library. A few things to consider:

- We have our own implementation of Aho-Corasic in DMS. That implementation was copied to SEER\*API. I think we should consider using an external library for this. https://github.com/robert-bor/aho-corasick looks like it would work.
- This library will be used externally as well as in SEER\*DMS. By default, the library will have the keyword definitions internally. That is the default usage pattern. When SEER*DMS uses it, we want to use the keyword definitions in the DMS database. We need to design this library to allow that.

As far as design the workflow will go

- initialization

   ```java
   // load default reportability keyword list
   var builder = new ReportabilityScreenerBuilder();

   // add internal keywords
   builder.defaultKeywords();

   // add external keywords one at a time
   builder.add(keyword);

   // or external keywords all at once
   builder.add(keywords);

   // finally call build
   var screener = builder.build();
   ```

- screen specific text

   ```java
   ScreeningResult result = screener.screen(text);
   ```

- the result will contain the following:

   ```java
   public class ScreeningResult {
      public enum ReportabilityResult {
         REPORTABLE, NON_REPORTABLE, UNKNOWN
      }

      ReportabilityResult result;

      List<Keyword> positiveKeywords;
      List<Keyword> negativeKeywords;
      List<KeyWord> otherKeywords;
   }
   ```

Note the `Keyword` class will include the keyword and the start/end position of the match as well as `ignored`.

As for the keywords themselves, I ran this query:

```sql
select g.name as group, count(*)
from screening_keyword sk
   join screening_group g on (sk.keyword_group = g.group_id)
where reportability_alg_keyword = true
group by 1;
```

and got the following. Do we need to keep the "Other" group? Do they factor into reportability?
            
REG | group | count
-- | -- | --
SE | Other | 378
SE | Negative | 625
SE | Positive | 1526

We will discuss in the meeting on Thursday.

### PR DESCRIPTION
**Pull request by garybeverungen**
Closes #1

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -10,7 +10,7 @@
 
     <properties>
         <!-- build under 21 but release as JDK 8 compatible -->
-        <java.version>21</java.version>
```
Comment (ctmay4) [inline pom.xml]: Why did you change this. It is correct. The `integration.yaml` is building using Java 21 but setting `--release` to Java 8.

```yaml
      - name: Set up JDK
        uses: actions/setup-java@v4
        with:
          distribution: 'adopt'
          java-version: '21'
          cache: 'maven'
```

File: src/main/java/com/imsweb/ReportabilityScreener.java [hunk 2]
```diff
@@ -1,16 +1,48 @@
...
+        result.setNegativeKeywords(_negativeTrie.parseText(text).stream().map(e -> new Keyword(e, Group.NEGATIVE)).collect(Collectors.toList()));
+        result.setOtherKeywords(_otherTrie.parseText(text).stream().map(e -> new Keyword(e, Group.OTHER)).collect(Collectors.toList()));
+
+        ignoreNegatedPositiveKeywordMatches(result.getPositiveKeywords(), result.getNegativeKeywords());
+        result.setResult(getResultBasedOnKeywordMatches(result.getPositiveKeywords()));
+
+        return result;
+    }
+
+    private void ignoreNegatedPositiveKeywordMatches(List<Keyword> positiveKeywords, List<Keyword> negativeKeywords) {
```
Comment (ctmay4) [inline src/main/java/com/imsweb/ReportabilityScreener.java]: Please add Javadoc explain method.

File: src/main/java/com/imsweb/ReportabilityScreener.java [hunk 3]
```diff
@@ -1,16 +1,48 @@
...
+        positiveKeywords.stream()
+                .filter(k -> negativeKeywords.stream().anyMatch(nk -> nk.getStart() <= k.getStart() && k.getEnd() <= nk.getEnd()))
+                .forEach(k -> k.setIgnored(true));
+    }
+
+    private ReportabilityResult getResultBasedOnKeywordMatches(List<Keyword> positiveKeywords) {
```
Comment (ctmay4) [inline src/main/java/com/imsweb/ReportabilityScreener.java]: Please add Javadoc explaining method

File: src/main/java/com/imsweb/ReportabilityScreenerBuilder.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    public void add(List<String> keywords, Group group) {
+        keywords.forEach(k -> add(k, group));
+    }
+
+    public void defaultKeywords() {
+        try (BufferedReader reader = new BufferedReader(new FileReader("src/main/resources/default.keyword.list.txt"))) {
+            reader.lines().map(l -> l.split("\\|")).forEach(l -> add(l[0], getGroupFromString(l[1])));
+        }
+        catch (IOException e) {
```
Comment (ctmay4) [inline src/main/java/com/imsweb/ReportabilityScreenerBuilder.java:65]: I actually think I would just convert this to an `IllegalStateException` and throw it. This should never happen, but if it does the library is useless.

File: src/main/java/com/imsweb/ReportabilityScreenerBuilder.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+import org.ahocorasick.trie.Trie;
+import org.ahocorasick.trie.Trie.TrieBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.imsweb.ReportabilityScreener.Group;
+
+public class ReportabilityScreenerBuilder {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ReportabilityScreener.class);
```
Comment (ctmay4) [inline src/main/java/com/imsweb/ReportabilityScreenerBuilder.java]: Sonarqube complains about this. You are using the wrong class for the logger.

File: src/main/java/com/imsweb/ReportabilityScreenerBuilder.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+                group = Group.POSITIVE;
+                break;
+            case "Negative":
+                group = Group.NEGATIVE;
+                break;
+            case "Other":
+                group = Group.OTHER;
+                break;
+            default:
+                throw new InvalidParameterException("Unexpected group value: " + groupString);
```
Comment (ctmay4) [inline src/main/java/com/imsweb/ReportabilityScreenerBuilder.java]: `InvalidParameterException` is from java.security. That doesn't seem right. Use `IllegalArgumentException` instead.

File: src/test/java/com/imsweb/ReportabilityScreenerTest.java [hunk 7]
```diff
@@ -10,8 +12,14 @@
 class ReportabilityScreenerTest {
 
     @Test
```
Comment (ctmay4) [inline src/test/java/com/imsweb/ReportabilityScreenerTest.java:21]: The tests need to be expanded. You need to have method-based tests for 

- ignoreNegatedPositiveKeywordMatches
- getResultBasedOnKeywordMatches

Also need to have tests that show initializing screening with external keywords.

**Review by ctmay4 — COMMENTED** [review]
Great start!

File: pom.xml (hunk 1 above)
Comment (garybeverungen) [reply pom.xml]: Sorry, I guess I misunderstood the emails about this. I thought the takeaway was that we had to build and release using the same java version. I changed it back.

File: src/main/java/com/imsweb/ReportabilityScreenerBuilder.java (hunk 5 above)
Comment (garybeverungen) [reply src/main/java/com/imsweb/ReportabilityScreenerBuilder.java]: Exceptions are no longer caught and logged, so I ended up removing the LOG altogether.

File: pom.xml (hunk 1 above)
Comment (ctmay4) [reply pom.xml]: The discussion was about language level. I was hoping we could use language features from later Java versions and still output Java 8 bytecode. No dice. We have to leave the language level at 8 in IntelliJ.

File: src/main/java/com/imsweb/ReportabilityScreenerBuilder.java (hunk 5 above)
Comment (ctmay4) [reply src/main/java/com/imsweb/ReportabilityScreenerBuilder.java]: Good call. We might want to remove all the logging libraries from the `pom.xml`. Let dependencies the better. Lets wait a bit longer to see if we will want to add logging somewhere else.

**Comment by ctmay4** [comment]
This is looking good. Can you just clean up the Sonar issues and I think this can be merged.

**Comment by garybeverungen** [comment]
SonarCloud is complaining about the use of `.collect(Collectors.toList()));`, but the `.toList()` is not allowed at this language level. I'm not sure how to resolve the issue.

**Comment by ctmay4** [comment]
Interesting. Sonar has the wrong language level. Let me see if I can fix that.

**Comment by ctmay4** [comment]
I tried to change a setting. If that doesn't work we can just switch the compiler to 8 and be done with it.

**Comment by ctmay4** [comment]
Nevermind. Looks like that fixed it!

File: src/main/java/com/imsweb/ReportabilityScreener.java [hunk 8]
```diff
@@ -1,16 +1,62 @@
...
+     * If there are any non-ignored positive keyword matches, the screened text is considered Reportable.
+     * Otherwise, it is considered Non-reportable.
+     * @param positiveKeywords positive keyword matches from the text being screened
+     * @return ReportabilityResult indicating whether the screened text is Reportable or Non-reportable.
+     */
+    protected ReportabilityResult getResultBasedOnKeywordMatches(List<Keyword> positiveKeywords) {
+        if (positiveKeywords.stream().anyMatch(k -> !k.isIgnored()))
+            return ReportabilityResult.REPORTABLE;
+        else
+            return ReportabilityResult.NON_REPORTABLE;
```
Comment (ctmay4) [inline src/main/java/com/imsweb/ReportabilityScreener.java]: Nitpicky but can you use static imports?

File: src/main/java/com/imsweb/ReportabilityScreenerBuilder.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+                group = Group.POSITIVE;
+                break;
+            case "Negative":
+                group = Group.NEGATIVE;
+                break;
+            case "Other":
+                group = Group.OTHER;
+                break;
+            default:
+                throw new IllegalArgumentException("Unexpected group value: " + groupString);
```
Comment (ctmay4) [inline src/main/java/com/imsweb/ReportabilityScreenerBuilder.java:107]: Code coverage shows you are not testing this.

File: src/main/java/com/imsweb/ReportabilityScreenerBuilder.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        Group group;
+        switch (groupString) {
+            case "Positive":
+                group = Group.POSITIVE;
+                break;
+            case "Negative":
+                group = Group.NEGATIVE;
+                break;
+            case "Other":
+                group = Group.OTHER;
```
Comment (ctmay4) [inline src/main/java/com/imsweb/ReportabilityScreenerBuilder.java]: Static imports please.

File: src/test/java/com/imsweb/ReportabilityScreenerTest.java [hunk 11]
```diff
@@ -10,8 +17,126 @@
...
+        List<Keyword> positiveKeywordMatches;
+        List<Keyword> negativeKeywordMatches;
+        ReportabilityScreener screener = new ReportabilityScreenerBuilder().build();
+
+        Keyword posKeyword10to20 = new Keyword("posKeyword", 10, 20, Group.POSITIVE);
+        Keyword posKeyword15to25 = new Keyword("posKeyword", 15, 25, Group.POSITIVE);
+        Keyword posKeyword15to16 = new Keyword("posKeyword", 15, 16, Group.POSITIVE);
+        Keyword negKeyword9to21 = new Keyword("negKeyword", 9, 21, Group.NEGATIVE);
+        Keyword negKeyword10to20 = new Keyword("negKeyword", 10, 20, Group.NEGATIVE);
+        Keyword negKeyword15to16 = new Keyword("negKeyword", 15, 16, Group.NEGATIVE);
```
Comment (ctmay4) [inline src/test/java/com/imsweb/ReportabilityScreenerTest.java]: More static imports.

**Comment by ctmay4** [comment]
Just some extremely minor things and then we can merge. Looks good.

File: src/main/java/com/imsweb/ReportabilityScreenerBuilder.java (hunk 9 above)
Comment (garybeverungen) [reply src/main/java/com/imsweb/ReportabilityScreenerBuilder.java:107]: Can this go in ReportabilityScreenerTest or should I make a new ReportabilityScreenerBuilderTest class?

Comment (ctmay4) [reply src/main/java/com/imsweb/ReportabilityScreenerBuilder.java:107]: I would create a new test class. That way any additions to the class could be tested there in the future.

File: src/main/java/com/imsweb/ReportabilityScreenerBuilder.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+public class ReportabilityScreenerBuilder {
+
+    private final TrieBuilder _positiveTrieBuilder;
+    private final TrieBuilder _negativeTrieBuilder;
+    private final TrieBuilder _otherTrieBuilder;
+
+    public ReportabilityScreenerBuilder() {
+        _positiveTrieBuilder = Trie.builder().onlyWholeWords();
+        _negativeTrieBuilder = Trie.builder().onlyWholeWords();
+        _otherTrieBuilder = Trie.builder().onlyWholeWords();
```
Comment (ctmay4) [inline src/main/java/com/imsweb/ReportabilityScreenerBuilder.java]: I think these are case-sensitive. I assume we want case-insensitive. Please make sure to add unit tests cases to sensitivity and whole words. I actually didn't think we wanted whole words only.

Comment (garybeverungen) [reply src/main/java/com/imsweb/ReportabilityScreenerBuilder.java]: Good catch. I don't know why I didn't think of that when I made the whole word only change. I'll change that and add unit tests. 

I'm pretty sure we do want whole-word-only. We have very few keywords in SEER*DMS that have whole_word_only=false, and none of them are reportability_alg_keywords.

**Comment by garybeverungen** [comment]
Should I add any keyword validation/cleaning to ReportabilityScreenerBuilder.add()?
For example, what should we do if a user tries to add the same keyword twice?
In SEER*DMS, keywords are a max of 200 characters. Should we enforce that?
Should we trim whitespace on the passed keyword?

**Comment by ctmay4** [comment]
> Should I add any keyword validation/cleaning to ReportabilityScreenerBuilder.add()? For example, what should we do if a user tries to add the same keyword twice? In SEER*DMS, keywords are a max of 200 characters. Should we enforce that? Should we trim whitespace on the passed keyword?

I'd say yes to all that!

File: src/main/java/com/imsweb/ReportabilityScreenerBuilder.java [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+                break;
+            case OTHER:
+                _otherTrieBuilder.addKeyword(keyword);
+                break;
+        }
+    }
+
+    public void add(List<String> keywords, Group group) {
+        keywords.forEach(k -> add(k, group));
+    }
```
Comment (ctmay4) [inline src/main/java/com/imsweb/ReportabilityScreenerBuilder.java:59]: I think another potentially useful thing to add to the builder would be `InputStream` and `Reader` add methods. Maybe we could do that in a separate issue if its too much here. The though process is someone could maintain their list in files and those could be loaded in a single command.

File: src/main/java/com/imsweb/ReportabilityScreenerBuilder.java [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+                group = POSITIVE;
+                break;
+            case "Negative":
+                group = NEGATIVE;
+                break;
+            case "Other":
+                group = OTHER;
+                break;
+            default:
+                throw new IllegalArgumentException("Unexpected group value: " + groupString);
```
Comment (ctmay4) [inline src/main/java/com/imsweb/ReportabilityScreenerBuilder.java]: Also, if people are going to be creating files at some point, we want to be a bit more lenient. This should match us comparing without case. So `POSITIVE` should also work. We might also want to support shorter versions like `POS` or even `P`. Just make it easier for people to maintain these files if they need to.

File: src/main/java/com/imsweb/ReportabilityScreenerBuilder.java (hunk 13 above)
Comment (ctmay4) [reply src/main/java/com/imsweb/ReportabilityScreenerBuilder.java:59]: Don’t bother with these new constructors for this issue, we can add it later if needed.

**Comment by ctmay4** [comment]
@garybeverungen I'd like to merge this soon since i need to work on the Maven Central release process. Do you want to finish this up first or I can merge this and we can create new issues for the few things left.

**Comment by garybeverungen** [comment]
I can add the extra validation and loosen the requirements for the getGroupFromString() method. That shouldn't take long. I'll try to have that done today. 
I think the `InputStream` and `Reader` methods can wait.

**Comment by ctmay4** [comment]
I agree that the stream and reader constructors are definitely a future update. Thanks.

**Comment by garybeverungen** [comment]
I added keyword validation and relaxed keyword group parsing. Let me know if there's anything else to do.

**Comment by ctmay4** [comment]
Looks good. Lets get this merged and I can work on getting the release to work.
